{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#The purpose of this is to use an exon centric analysis:\n",
    "#Take all exons and label as constitutive vs spliced \n",
    "#Associate exons with iCLIP binding to get the total"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/sam/anaconda/lib/python2.7/site-packages/IPython/html.py:14: ShimWarning: The `IPython.html` package has been deprecated. You should import from `notebook` instead. `IPython.html.widgets` has moved to `ipywidgets`.\n",
      "  \"`IPython.html.widgets` has moved to `ipywidgets`.\", ShimWarning)\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "pd.options.mode.chained_assignment = None  # default='warn'\n",
    "import matplotlib as mpl\n",
    "import matplotlib.pyplot as plt\n",
    "import os\n",
    "from math import *\n",
    "from subprocess import *\n",
    "import pybedtools as pbt\n",
    "from glob import glob\n",
    "import seaborn as sns\n",
    "import statsmodels.formula.api as sm\n",
    "import csv\n",
    "import re\n",
    "csv.register_dialect(\"textdialect\",delimiter='\\t')\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#datasets to use for RT stops = mega dataframe_gene_expression_normalized_1nt\n",
    "#this one has all of the information and is also normalized with gene calls to the highest expression in a given sample\n",
    "clip_df = pd.read_csv('hnM_DMSO_TAM_master_dataframe_gene_expression_normalized_1nt',sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cluster_id</th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>ens_gene</th>\n",
       "      <th>blank</th>\n",
       "      <th>strand</th>\n",
       "      <th>hnM_DMSO</th>\n",
       "      <th>hnM_TAM</th>\n",
       "      <th>hnM_DMSO_1</th>\n",
       "      <th>...</th>\n",
       "      <th>external_gene_name</th>\n",
       "      <th>biotype</th>\n",
       "      <th>twist_d0_fpkm_1</th>\n",
       "      <th>twist_d14_fpkm_1</th>\n",
       "      <th>log2ratio_fpkm_d14_d0</th>\n",
       "      <th>hnM_DMSO_clip_gene_fpkm_norm</th>\n",
       "      <th>hnM_TAM_clip_gene_fpkm_norm</th>\n",
       "      <th>hnM_DMSO_clip_gene_fpkm_norm_1</th>\n",
       "      <th>hnM_TAM_clip_gene_fpkm_norm_1</th>\n",
       "      <th>log2ratio_hnM_TAM_DMSO_clip_gene_fpkm_norm</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1:251639-251640</td>\n",
       "      <td>chr1</td>\n",
       "      <td>251639</td>\n",
       "      <td>251640</td>\n",
       "      <td>ENSG00000228463</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>6.521757</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>7.521757</td>\n",
       "      <td>...</td>\n",
       "      <td>AP006222.2</td>\n",
       "      <td>lincRNA</td>\n",
       "      <td>1.040763</td>\n",
       "      <td>1.117389</td>\n",
       "      <td>0.102490</td>\n",
       "      <td>6.266323</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>7.266323</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>-2.861225</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1:251646-251647</td>\n",
       "      <td>chr1</td>\n",
       "      <td>251646</td>\n",
       "      <td>251647</td>\n",
       "      <td>ENSG00000228463</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>2.898558</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3.898558</td>\n",
       "      <td>...</td>\n",
       "      <td>AP006222.2</td>\n",
       "      <td>lincRNA</td>\n",
       "      <td>1.040763</td>\n",
       "      <td>1.117389</td>\n",
       "      <td>0.102490</td>\n",
       "      <td>2.785032</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3.785032</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>-1.920306</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1:564514-564515</td>\n",
       "      <td>chr1</td>\n",
       "      <td>564514</td>\n",
       "      <td>564515</td>\n",
       "      <td>ENSG00000225972</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>5.691002</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>MTND1P23</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>9.666940</td>\n",
       "      <td>27.279600</td>\n",
       "      <td>1.496691</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.208618</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.208618</td>\n",
       "      <td>0.273358</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1:564595-564596</td>\n",
       "      <td>chr1</td>\n",
       "      <td>564595</td>\n",
       "      <td>564596</td>\n",
       "      <td>ENSG00000225972</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>5.072477</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>6.072477</td>\n",
       "      <td>...</td>\n",
       "      <td>MTND1P23</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>9.666940</td>\n",
       "      <td>27.279600</td>\n",
       "      <td>1.496691</td>\n",
       "      <td>0.524724</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.524724</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>-0.608548</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1:564722-564723</td>\n",
       "      <td>chr1</td>\n",
       "      <td>564722</td>\n",
       "      <td>564723</td>\n",
       "      <td>ENSG00000225972</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>19.349407</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>MTND1P23</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>9.666940</td>\n",
       "      <td>27.279600</td>\n",
       "      <td>1.496691</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.709300</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.709300</td>\n",
       "      <td>0.773405</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 32 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           cluster_id chrom   start     end         ens_gene  blank strand  \\\n",
       "0  chr1:251639-251640  chr1  251639  251640  ENSG00000228463      0      -   \n",
       "1  chr1:251646-251647  chr1  251646  251647  ENSG00000228463      0      -   \n",
       "2  chr1:564514-564515  chr1  564514  564515  ENSG00000225972      0      +   \n",
       "3  chr1:564595-564596  chr1  564595  564596  ENSG00000225972      0      +   \n",
       "4  chr1:564722-564723  chr1  564722  564723  ENSG00000225972      0      +   \n",
       "\n",
       "   hnM_DMSO    hnM_TAM  hnM_DMSO_1                    ...                      \\\n",
       "0  6.521757   0.000000    7.521757                    ...                       \n",
       "1  2.898558   0.000000    3.898558                    ...                       \n",
       "2  0.000000   5.691002    1.000000                    ...                       \n",
       "3  5.072477   0.000000    6.072477                    ...                       \n",
       "4  0.000000  19.349407    1.000000                    ...                       \n",
       "\n",
       "   external_gene_name                 biotype twist_d0_fpkm_1  \\\n",
       "0          AP006222.2                 lincRNA        1.040763   \n",
       "1          AP006222.2                 lincRNA        1.040763   \n",
       "2            MTND1P23  unprocessed_pseudogene        9.666940   \n",
       "3            MTND1P23  unprocessed_pseudogene        9.666940   \n",
       "4            MTND1P23  unprocessed_pseudogene        9.666940   \n",
       "\n",
       "  twist_d14_fpkm_1  log2ratio_fpkm_d14_d0  hnM_DMSO_clip_gene_fpkm_norm  \\\n",
       "0         1.117389               0.102490                      6.266323   \n",
       "1         1.117389               0.102490                      2.785032   \n",
       "2        27.279600               1.496691                      0.000000   \n",
       "3        27.279600               1.496691                      0.524724   \n",
       "4        27.279600               1.496691                      0.000000   \n",
       "\n",
       "   hnM_TAM_clip_gene_fpkm_norm  hnM_DMSO_clip_gene_fpkm_norm_1  \\\n",
       "0                     0.000000                        7.266323   \n",
       "1                     0.000000                        3.785032   \n",
       "2                     0.208618                        1.000000   \n",
       "3                     0.000000                        1.524724   \n",
       "4                     0.709300                        1.000000   \n",
       "\n",
       "   hnM_TAM_clip_gene_fpkm_norm_1 log2ratio_hnM_TAM_DMSO_clip_gene_fpkm_norm  \n",
       "0                       1.000000                                  -2.861225  \n",
       "1                       1.000000                                  -1.920306  \n",
       "2                       1.208618                                   0.273358  \n",
       "3                       1.000000                                  -0.608548  \n",
       "4                       1.709300                                   0.773405  \n",
       "\n",
       "[5 rows x 32 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clip_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['cluster_id',\n",
       " 'chrom',\n",
       " 'start',\n",
       " 'end',\n",
       " 'ens_gene',\n",
       " 'blank',\n",
       " 'strand',\n",
       " 'hnM_DMSO',\n",
       " 'hnM_TAM',\n",
       " 'hnM_DMSO_1',\n",
       " 'hnM_TAM_1',\n",
       " 'log2ratio_TAM-DMSO',\n",
       " 'gene',\n",
       " 'gene_coords',\n",
       " 'twist_d0_fpkm',\n",
       " 'twist_d14_fpkm',\n",
       " 'twist_d0_d14_fpkm_sum',\n",
       " 'twist_d0_tpm',\n",
       " 'twist_d14_tpm',\n",
       " 'chr',\n",
       " 'coords',\n",
       " 'gene_length',\n",
       " 'external_gene_name',\n",
       " 'biotype',\n",
       " 'twist_d0_fpkm_1',\n",
       " 'twist_d14_fpkm_1',\n",
       " 'log2ratio_fpkm_d14_d0',\n",
       " 'hnM_DMSO_clip_gene_fpkm_norm',\n",
       " 'hnM_TAM_clip_gene_fpkm_norm',\n",
       " 'hnM_DMSO_clip_gene_fpkm_norm_1',\n",
       " 'hnM_TAM_clip_gene_fpkm_norm_1',\n",
       " 'log2ratio_hnM_TAM_DMSO_clip_gene_fpkm_norm']"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clip_df.columns.tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#Splicing library\n",
    "#Description of ucsc known alt\n",
    "# This track shows various types of alternative splicing and other events that result in more than a single transcript from the same gene. The label by an item describes the type of event. The events are:\n",
    "\n",
    "# Alternate Promoter (altPromoter) - Transcription starts at multiple places. The altPromoter extends from 100 bases before to 50 bases after transcription start.\n",
    "# Alternate Finish Site (altFinish) - Transcription ends at multiple places.\n",
    "# Cassette Exon (cassetteExon) - Exon is present in some transcripts but not others. These are found by looking for exons that overlap an intron in the same transcript.\n",
    "# Retained Intron (retainedIntron) - Introns are spliced out in some transcripts but not others. In some cases, particularly when the intron is near the 3' end, this can reflect an incompletely processed transcript rather than a true alt-splicing event.\n",
    "# Overlapping Exon (bleedingExon) - Initial or terminal exons overlap in an intron in another transcript. These often are associated with incompletely processed transcripts.\n",
    "# Alternate 3' End (altThreePrime) - Variations on the 3' end of an intron.\n",
    "# Alternate 5' End (altFivePrime) - Variations on the 5' end of an intron.\n",
    "# Intron Ends have AT/AC (atacIntron) - An intron with AT/AC ends rather than the usual GT/AG. These are associated with the minor spliceosome.\n",
    "# Strange Intron Ends (strangeSplice) - An intron with ends that are not GT/AG, GC/AG, or AT/AC. These are usually artifacts of some sort due to sequencing error or polymorphism\n",
    "\n",
    "#rmats splicing library\n",
    "def getASEvents(asDir,): ## get AS events from GTF and SAM files\n",
    "  logging.debug(\"getting AS events function..\");\n",
    "\n",
    "  samNames1=uniqSamNames(1);\n",
    "  samNames2=uniqSamNames(2);\n",
    "  finalSamNames = ','.join(samNames1)+','+','.join(samNames2);  \n",
    "  if lite==1:\n",
    "    finalSamNames = ','.join(samNames1); ## for the lite version\n",
    "\n",
    "  #cmd = 'python '+binPath+'/processGTF.SAMs.py '+gtf+' '+asDir+'/fromGTF'+' '+finalSamNames+' '+libType+' '+tempPath;\n",
    "  cmd = 'python '+binPath+'/processGTF.BAMs.py '+gtf+' '+asDir+'/fromGTF'+' '+finalSamNames+' '+libType+' '+SEPE+' '+str(novelSS) +' '+tempPath;\n",
    "  #oFile.write('###### getting AS events from GTF and SAM files #####\\n'+cmd+'\\n#\\n');\n",
    "  oFile.write('###### getting AS events from GTF and BAM files #####\\n'+cmd+'\\n#\\n');\n",
    "  oFile.flush();\n",
    "  status,output=commands.getstatusoutput(cmd);\n",
    "  logging.debug(\"getting AS events is done with status %s\" % status);\n",
    "  if (int(status)!=0): ## it did not go well\n",
    "    logging.debug(\"error in getting AS events %s\" % status);\n",
    "    logging.debug(\"error detail: %s\" % output);\n",
    "    raise Exception();\n",
    "  logging.debug(output);"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#collect a dataframe with only the information we need for each sample\n",
    "dmso_names = ['chrom','start','end','ens_gene','hnM_DMSO_clip_gene_fpkm_norm']\n",
    "dmso = "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#add an id to each exon\n",
    "with open('spliced_exon_analysis/hg19_v19_exons.bed') as f:\n",
    "    counter = 1\n",
    "    outlist = []\n",
    "    for line in f:\n",
    "        line = line.strip().split('\\t')\n",
    "        line[4] = \"exon_\"+str(counter)\n",
    "        counter += 1\n",
    "        outlist.append(line)\n",
    "with open('spliced_exon_analysis/hg19_v19_exons_with_exon_id.bed','w') as f:\n",
    "    writer = csv.writer(f,'textdialect')\n",
    "    writer.writerows(outlist)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(38.18791573937649, 'percent of exons are alternative')\n"
     ]
    }
   ],
   "source": [
    "#first step collect exons and associate with features\n",
    "hg19_v19_exons = pd.read_csv('spliced_exon_analysis/hg19_v19_exons_with_exon_id.bed',sep='\\t',names=['chrom','start','stop','ens_gene','exon_id','strand'])\n",
    "#remove gene name stuff after ens_gene\n",
    "hg19_v19_exons['ens_gene']=hg19_v19_exons.ens_gene.str.split('.').str[0]    \n",
    "#read in gene names and biotypes\n",
    "#I got this from ensembl biomart using the last GRCh37 archive - ENS version 75\n",
    "hg19_grch37_gene_names_and_biotype = pd.read_csv('spliced_exon_analysis/ensembl_ens_gene_gene_name_biotype_grch37_p13',sep='\\t',names=['ens_gene','external_gene_name','biotype'])\n",
    "#why are there duplicates?\n",
    "hg19_grch37_gene_names_and_biotype.drop_duplicates(inplace=True)\n",
    "\n",
    "#merge the exons with the gene names and biotypes\n",
    "hg19_v19_exons = pd.merge(hg19_v19_exons,hg19_grch37_gene_names_and_biotype,on='ens_gene',how='left')\n",
    "hg19_v19_exons.to_csv('spliced_exon_analysis/hg19_v19_exon_gene_names_and_biotypes',sep='\\t',header=False,index=False,na_rep='NaN')\n",
    "\n",
    "#create a bedtools\n",
    "hg19_v19_exons_bt = pbt.BedTool.from_dataframe(hg19_v19_exons,na_rep='NaN')\n",
    "\n",
    "#now I need to bring in the library of splice exons\n",
    "#Lets try multiple sources - first use hg38 known alt\n",
    "#first - hg38 known alt but lifted to hg19 - this has many more events - roughly double the number of cassette exons\n",
    "known_alt = pd.read_csv('/media/sam/Data1/annotations/splicing_libraries/knownAlt_hg38_to_hg19.bed',sep='\\t',names=['chrom','start','end','splicing_type','score','strand'])\n",
    "known_alt_bt = pbt.BedTool.from_dataframe(known_alt)\n",
    "known_alt_cassette = known_alt[known_alt.splicing_type == \"cassetteExon\"]\n",
    "known_alt_cassette_bt = pbt.BedTool.from_dataframe(known_alt_cassette).sort()\n",
    "\n",
    "#intersect all of known alt with the hg19_v19_exons to find those that don't intersect with any known splicing event and are thus constitutive. also collect those that are alternative\n",
    "hg19_v19_constitutive_exons = hg19_v19_exons_bt.intersect(b=known_alt_bt,s=True,v=True,wa=True)\n",
    "hg19_v19_alternative_exons = hg19_v19_exons_bt.intersect(b=known_alt_bt,s=True,u=True,wa=True)\n",
    "print((float(len(hg19_v19_alternative_exons))/float(len(hg19_v19_exons_bt))*100),'percent of exons are alternative')\n",
    "\n",
    "#now figure out which alternative exons are cassette exons. intersect alternative with cassette \n",
    "#should i require a fraction of overlap? \n",
    "#should I require there be a perfect match or a partial match?\n",
    "hg19_v19_alternative_cassette_exons = hg19_v19_alternative_exons.intersect(b=known_alt_cassette_bt,wa=True,s=True,u=True,f=1.0,r=True)\n",
    "hg19_v19_alternative_non_cassette_exons = hg19_v19_alternative_exons.intersect(b=known_alt_cassette_bt,wa=True,s=True,v=True,f=1.0,r=True)\n",
    "\n",
    "#focus on cassette exons to start\n",
    "#in the dataframe label exons as alternative or not alternative and then cassette or non-cassette\n",
    "# hg19_v19_alternative_exons.to_dataframe\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#now I need to bring in the library of splice exons\n",
    "#Lets try multiple sources - first use hg38 known alt\n",
    "#first - hg38 known alt but lifted to hg19 - this has many more events - roughly double the number of cassette exons\n",
    "known_alt = pd.read_csv('/media/sam/Data1/annotations/splicing_libraries/knownAlt_hg38_to_hg19.bed',sep='\\t',names=['chrom','start','end','splicing_type','score','strand'])\n",
    "known_alt_bt = pbt.BedTool.from_dataframe(known_alt)\n",
    "known_alt_cassette = known_alt[known_alt.splicing_type == \"cassetteExon\"]\n",
    "known_alt_cassette_bt = pbt.BedTool.from_dataframe(known_alt_cassette).sort()\n",
    "\n",
    "#intersect all of known alt with the hg19_v19_exons to find those that don't intersect with any known splicing event and are thus constitutive. also collect those that are alternative\n",
    "hg19_v19_constitutive_exons = hg19_v19_exons_bt.intersect(b=known_alt_bt,s=True,v=True,wa=True)\n",
    "hg19_v19_alternative_exons = hg19_v19_exons_bt.intersect(b=known_alt_bt,s=True,u=True,wa=True)\n",
    "print((float(len(hg19_v19_alternative_exons))/float(len(hg19_v19_exons_bt))*100),'percent of exons are alternative')\n",
    "\n",
    "#now figure out which alternative exons are cassette exons. intersect alternative with cassette \n",
    "#should i require a fraction of overlap? \n",
    "#should I require there be a perfect match or a partial match?\n",
    "hg19_v19_alternative_cassette_exons = hg19_v19_alternative_exons.intersect(b=known_alt_cassette_bt,wa=True,s=True,u=True,f=1.0,r=True)\n",
    "hg19_v19_alternative_non_cassette_exons = hg19_v19_alternative_exons.intersect(b=known_alt_cassette_bt,wa=True,s=True,v=True,f=1.0,r=True)\n",
    "\n",
    "#focus on cassette exons to start\n",
    "#in the dataframe label exons as alternative or not alternative and then cassette or non-cassette\n",
    "# hg19_v19_alternative_exons.to_dataframe\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "123890\n",
      "123890\n"
     ]
    }
   ],
   "source": [
    "print(len(hg19_v19_alternative_cassette_exons)+len(hg19_v19_alternative_non_cassette_exons))\n",
    "print(len(hg19_v19_alternative_exons))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>splicing_type</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>altFinish</th>\n",
       "      <td>16198</td>\n",
       "      <td>16198</td>\n",
       "      <td>16198</td>\n",
       "      <td>16198</td>\n",
       "      <td>16198</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>altFivePrime</th>\n",
       "      <td>8305</td>\n",
       "      <td>8305</td>\n",
       "      <td>8305</td>\n",
       "      <td>8305</td>\n",
       "      <td>8305</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>altPromoter</th>\n",
       "      <td>68284</td>\n",
       "      <td>68284</td>\n",
       "      <td>68284</td>\n",
       "      <td>68284</td>\n",
       "      <td>68284</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>altThreePrime</th>\n",
       "      <td>14129</td>\n",
       "      <td>14129</td>\n",
       "      <td>14129</td>\n",
       "      <td>14129</td>\n",
       "      <td>14129</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>atacIntron</th>\n",
       "      <td>276</td>\n",
       "      <td>276</td>\n",
       "      <td>276</td>\n",
       "      <td>276</td>\n",
       "      <td>276</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>bleedingExon</th>\n",
       "      <td>41343</td>\n",
       "      <td>41343</td>\n",
       "      <td>41343</td>\n",
       "      <td>41343</td>\n",
       "      <td>41343</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cassetteExon</th>\n",
       "      <td>64926</td>\n",
       "      <td>64926</td>\n",
       "      <td>64926</td>\n",
       "      <td>64926</td>\n",
       "      <td>64926</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>retainedIntron</th>\n",
       "      <td>24924</td>\n",
       "      <td>24924</td>\n",
       "      <td>24924</td>\n",
       "      <td>24924</td>\n",
       "      <td>24924</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>strangeSplice</th>\n",
       "      <td>6130</td>\n",
       "      <td>6130</td>\n",
       "      <td>6130</td>\n",
       "      <td>6130</td>\n",
       "      <td>6130</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                chrom  start    end  score  strand\n",
       "splicing_type                                     \n",
       "altFinish       16198  16198  16198  16198   16198\n",
       "altFivePrime     8305   8305   8305   8305    8305\n",
       "altPromoter     68284  68284  68284  68284   68284\n",
       "altThreePrime   14129  14129  14129  14129   14129\n",
       "atacIntron        276    276    276    276     276\n",
       "bleedingExon    41343  41343  41343  41343   41343\n",
       "cassetteExon    64926  64926  64926  64926   64926\n",
       "retainedIntron  24924  24924  24924  24924   24924\n",
       "strangeSplice    6130   6130   6130   6130    6130"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "splicing_type_group = known_alt.groupby(by='splicing_type')\n",
    "splicing_type_group.count()\n",
    "print()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>2.445150e+05</td>\n",
       "      <td>2.445150e+05</td>\n",
       "      <td>244515.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>7.175108e+07</td>\n",
       "      <td>7.175145e+07</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>5.560061e+07</td>\n",
       "      <td>5.560064e+07</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>9.110000e+02</td>\n",
       "      <td>3.048000e+03</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>3.125139e+07</td>\n",
       "      <td>3.125145e+07</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>5.533152e+07</td>\n",
       "      <td>5.533339e+07</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>1.052106e+08</td>\n",
       "      <td>1.052108e+08</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>2.492005e+08</td>\n",
       "      <td>2.492006e+08</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              start           end     score\n",
       "count  2.445150e+05  2.445150e+05  244515.0\n",
       "mean   7.175108e+07  7.175145e+07       0.0\n",
       "std    5.560061e+07  5.560064e+07       0.0\n",
       "min    9.110000e+02  3.048000e+03       0.0\n",
       "25%    3.125139e+07  3.125145e+07       0.0\n",
       "50%    5.533152e+07  5.533339e+07       0.0\n",
       "75%    1.052106e+08  1.052108e+08       0.0\n",
       "max    2.492005e+08  2.492006e+08       0.0"
      ]
     },
     "execution_count": 79,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "known_alt.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>3.244220e+05</td>\n",
       "      <td>3.244220e+05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>7.513365e+07</td>\n",
       "      <td>7.513404e+07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>5.599167e+07</td>\n",
       "      <td>5.599167e+07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>5.770000e+02</td>\n",
       "      <td>6.470000e+02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>3.136982e+07</td>\n",
       "      <td>3.136999e+07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>6.143854e+07</td>\n",
       "      <td>6.143943e+07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>1.113556e+08</td>\n",
       "      <td>1.113557e+08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>2.492308e+08</td>\n",
       "      <td>2.492312e+08</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              start          stop\n",
       "count  3.244220e+05  3.244220e+05\n",
       "mean   7.513365e+07  7.513404e+07\n",
       "std    5.599167e+07  5.599167e+07\n",
       "min    5.770000e+02  6.470000e+02\n",
       "25%    3.136982e+07  3.136999e+07\n",
       "50%    6.143854e+07  6.143943e+07\n",
       "75%    1.113556e+08  1.113557e+08\n",
       "max    2.492308e+08  2.492312e+08"
      ]
     },
     "execution_count": 80,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hg19_v19_exons.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr1\t120720\t120932\tcassetteExon\t0\t-\n",
      " chr1\t120873\t120932\tcassetteExon\t0\t-\n",
      " chr1\t231300\t231385\tcassetteExon\t0\t-\n",
      " chr1\t543334\t543436\tcassetteExon\t0\t-\n",
      " chr1\t563340\t563603\tcassetteExon\t0\t-\n",
      " chr1\t634306\t634339\tcassetteExon\t0\t-\n",
      " chr1\t639064\t639169\tcassetteExon\t0\t-\n",
      " chr1\t647090\t647302\tcassetteExon\t0\t-\n",
      " chr1\t647243\t647302\tcassetteExon\t0\t-\n",
      " chr1\t655411\t655580\tcassetteExon\t0\t-\n",
      " "
     ]
    }
   ],
   "source": [
    "known_alt_cassette_bt.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#Old way I tried this\n",
    "\n",
    "\n",
    "#NOTE - THIS IS NOT THE BEST WAY TO WORK ON THIS... BETTER TO DO AN EXON CENTRIC OR SPLICING EVENT CENTRIC ANALYSIS\n",
    "\n",
    "#NOW FIGURE OUT HOW TO BRING IN THE ALTERNATIVE EXONS\n",
    "#put this in the directory spliced_exon_analysis\n",
    "\n",
    "#Goal - associate the clusters with nearest exon using bedtools closest\n",
    "\n",
    "#Ideal setup - have each cluster associated with nearest 1 exon. This exon must be able to be labeled as constitutive vs. alternative. This exon must then be able to be associated with a PSI value from each of our RNA seq analyses... How can I do this? Many of the splicing events will not match exactly.\n",
    "\n",
    "#Associate clusters with all exons first. Mark whether the exon is constitutive or variable.\n",
    "\n",
    "#HOW CAN I KNOW WHETHER AN EXON IS CONSTITUTIVE OR ALTERNATIVE?\n",
    "#WHY NOT USE RMATS FOR TWIST-ER CELLS? All SE called by rMATS will be merged and take the longest one. If an exon overlaps with this list, it is alternative. If it does not, it is constitutive\n",
    "#Caveat with this - We don't want false positives. maxPSI > 15%, minPSI < 85% ensures we don't take all skipped or all included exons.\n",
    "#Another source of information - MISO splicing library\n",
    "#Another source = hg19 knownAlt\n",
    "#If the exon is going to be paired with a PSI value it must be alternative\n",
    "\n",
    "#Then associate clusters with alternative exons from rMATS. Presumably rMATS exons are all alternatively spliced and contain no constitutively spliced exons. Can separate significantly differentially spliced from control using Carsten's publication \"control alternative\" cutoff (FDR > 50%, maxPSI of > 15%, minPSI < 85%)\n",
    "\n",
    "#As an exon annotation, I am using the hg19_v19 exons from CLIPPER, this is also copied in this directory\n",
    "#Other exon annotations that are available:\n",
    "#FAST-iCLIP-docs = hg19_transcriptome_collapse_exon.bed\n",
    "#Extract transcript regions script\n",
    "hg19_v19_exons = pd.read_csv('spliced_exon_analysis/hg19_v19_exons.bed',sep='\\t',names=['chrom','start','stop','ens_gene','blank','strand'])\n",
    "hg19_v19_exons['ens_gene']=hg19_v19_exons.ens_gene.str.split('.').str[0]\n",
    "#read in gene names and biotypes\n",
    "#I got this from ensembl biomart using the last GRCh37 archive - ENS version 75\n",
    "hg19_grch37_gene_names_and_biotype = pd.read_csv('spliced_exon_analysis/ensembl_ens_gene_gene_name_biotype_grch37_p13',sep='\\t',names=['ens_gene','external_gene_name','biotype'])\n",
    "#why are there duplicates?\n",
    "hg19_grch37_gene_names_and_biotype.drop_duplicates(inplace=True)\n",
    "\n",
    "#merge the exons with the gene names and biotypes\n",
    "hg19_v19_exons = pd.merge(hg19_v19_exons,hg19_grch37_gene_names_and_biotype,on='ens_gene',how='left')\n",
    "hg19_v19_exons.to_csv('spliced_exon_analysis/hg19_v19_exon_gene_names_and_biotypes',sep='\\t',header=False,index=False,na_rep='NaN')\n",
    "\n",
    "#save a copy of the merge_df clusters as a bedtools\n",
    "cluster_column_list = ['chrom','start','end','ens_gene','cluster_id','strand']\n",
    "merge_df.to_csv('spliced_exon_analysis/merge_df.bed',sep='\\t',na_rep='NaN',columns=cluster_column_list,header=False,index=False)\n",
    "\n",
    "#run closest bed with exons and keep the distance as a column\n",
    "#handle ties by taking both items\n",
    "#This means that some of the clusters will be associated with multiple exons\n",
    "!bedtools closest -a spliced_exon_analysis/merge_df.bed -b spliced_exon_analysis/hg19_v19_exon_gene_names_and_biotypes -D \"ref\" -s -t \"all\" > spliced_exon_analysis/merge_df_closest_exons\n",
    "\n",
    "!bedtools closest -a spliced_exon_analysis/merge_df.bed -b spliced_exon_analysis/hg19_v19_exon_gene_names_and_biotypes -D \"ref\" -s -t \"first\" > spliced_exon_analysis/merge_df_closest_exons_first_ties\n",
    "\n",
    "#bring in the data from TwistER, CC, and HH rMATS into splicing analysis folder\n",
    "#I used parse_rMATS_SE -fdr 0.05 -psi 0.1 -read 20\n",
    "#NOTE - it may be better for me in the future to just read in the data and parse it myself however I'd like - do this later\n",
    "\n",
    "#Should I do bedtools closest on each cluster next to an exon from CC, HH, TwistER? Set a limit of maximal distance\n",
    "\n",
    "#bring in the data from TwistER, CC, and HH rMATS into splicing analysis folder\n",
    "#I used parse_rMATS_SE -fdr 0.05 -psi 0.1 -read 20\n",
    "#NOTE - it may be better for me in the future to just read in the data and parse it myself however I'd like - do this later\n",
    "#Should I do bedtools closest on each cluster next to an exon from CC, HH, TwistER? Set a limit of maximal distance\n",
    "cc1_cc2_rMATS_df = pd.read_csv('spliced_exon_analysis/CC-1_CC-2_rMATS_3_2_5_SE_ROT_FDR_0.05_dPSI_0.1_read_cutoff_20.txt',sep='\\t')\n",
    "cc3_cc4_rMATS_df = pd.read_csv('spliced_exon_analysis/CC-3_CC-4_rMATS_3_2_5_SE_ROT_FDR_0.05_dPSI_0.1_read_cutoff_20.txt',sep='\\t')\n",
    "hh5_hh6_rMATS_df = pd.read_csv('spliced_exon_analysis/HH-5_HH-6_rMATS_3_2_5_SE_ROT_FDR_0.05_dPSI_0.1_read_cutoff_20.txt',sep='\\t')\n",
    "twist_d0_d14_rMATS_df = pd.read_csv('spliced_exon_analysis/SEH-1_SEH-2_rMATS_3_2_5_SE_ROT_FDR_0.05_dPSI_0.1_read_cutoff_20.txt',sep='\\t')\n",
    "\n",
    "#Now that I have this, I want to do bedtools closest to see if this this good"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
