#!/bin/bash

#This script will run CLIPPER on the file from Ryan's FAST-iCLIP pipeline using the --premrna option as well as superlocal to compare with
#the clipper clusters generated by Ryan's original pipeline

#Ryan original command cmd4 = "clipper --bam {} {} --outfile={} > /dev/null 2>&1".format(bamfile_sorted, genome, CLIPPERout_dup)
#bamfile_sorted = *_threshold=2_hg19_allreads.mergedRT_snoRNAremoved_miRNAremoved.srt.bam

#Note I renamed the file to take out the = sign in case it was causing problems

#preprocessing from CLIPPER website - just need to index the bam file
#samtools flagstat /media/sam/Data1/hnRNPM_iCLIP/iCLIP_analysis/2016_programmatic_iCLIP_analysis/getting_clipper_to_work/clipper_runs/hnM_DMSO_mergedRT_snoRNAremoved_miRNAremoved_sorted.bam

#samtools index /media/sam/Data1/hnRNPM_iCLIP/iCLIP_analysis/2016_programmatic_iCLIP_analysis/getting_clipper_to_work/clipper_runs/hnM_TAM_mergedRT_snoRNAremoved_miRNAremoved_sorted.bam

#Try with original command - failed... I have to add the bonferroni flag or clipper will crash
#It seems that in peakfinder.py, the bonferroni correction isn't even happening... so just include the flag...
#clipper -v -b /media/sam/Data1/hnRNPM_iCLIP/iCLIP_analysis/2016_programmatic_iCLIP_analysis/getting_clipper_to_work/clipper_runs/hnM_TAM_mergedRT_snoRNAremoved_miRNAremoved_sorted.bam -s hg19 -o clipper_out --bonferroni

#Try with --premRNA but not superlocal
#clipper -v -b /media/sam/Data1/hnRNPM_iCLIP/iCLIP_analysis/2016_programmatic_iCLIP_analysis/getting_clipper_to_work/clipper_runs/hnM_TAM_mergedRT_snoRNAremoved_miRNAremoved_sorted.bam -s hg19 -o clipper_out_premrna --bonferroni --premRNA

#Try with --superlocal but not --premRNA
#clipper -v -b /media/sam/Data1/hnRNPM_iCLIP/iCLIP_analysis/2016_programmatic_iCLIP_analysis/getting_clipper_to_work/clipper_runs/hnM_TAM_mergedRT_snoRNAremoved_miRNAremoved_sorted.bam -s hg19 -o clipper_out_superlocal --bonferroni --superlocal

#Try with --premRNA and superlocal
#clipper -v -b /media/sam/Data1/hnRNPM_iCLIP/iCLIP_analysis/2016_programmatic_iCLIP_analysis/getting_clipper_to_work/clipper_runs/hnM_TAM_clipper/hnM_TAM_mergedRT_snoRNAremoved_miRNAremoved_sorted.bam -s hg19 -o hnM_TAM_clipper/clipper_out_premrna_superlocal --bonferroni --superlocal --premRNA

#--------------------------
#Now with yeo lab parameters per Gabriel Pratt
#--bonferroni --superlocal --threshold-method binomial

#Try with yeo lab parameters - they just add a --threshold-method binomial
#REMEMBER TO ADD DUPL TO THE NAME - THERE ARE MULTIPLE GENES COUNTED MULTIPLE TIMES
#DMSO
clipper -v -b /media/sam/Data1/hnRNPM_iCLIP/iCLIP_analysis/2016_programmatic_iCLIP_analysis/getting_clipper_to_work/clipper_runs/hnM_DMSO_clipper/hnM_DMSO_mergedRT_snoRNAremoved_miRNAremoved_sorted.bam -s hg19 -o hnM_DMSO_clipper/clipper_out_premrna_superlocal_threshold_binomial_dupl --bonferroni --superlocal --premRNA --threshold-method binomial
#TAM
#clipper -v -b /media/sam/Data1/hnRNPM_iCLIP/iCLIP_analysis/2016_programmatic_iCLIP_analysis/getting_clipper_to_work/clipper_runs/hnM_TAM_clipper/hnM_TAM_mergedRT_snoRNAremoved_miRNAremoved_sorted.bam -s hg19 -o hnM_TAM_clipper/clipper_out_premrna_superlocal_threshold_binomial --bonferroni --superlocal --premRNA --threshold-method binomial
#-------------------------

#current error message - there are some error messages when running CLIPPER - how to deal with this
#May be appropriate to ignore - there are not that many of them anyway
#For example, the yeo lab parameter clipper run too 2 hours and only generated 7 of these errors

#WARNING:py.warnings:/home/sam/anaconda/lib/python2.7/site-packages/scipy/interpolate/fitpack2.py:222: UserWarning: 
#The maximal number of iterations maxit (set to 20 by the program)
#allowed for finding a smoothing spline with fp=s has been reached: s
#too small.
#There is an approximation returned but the corresponding weighted sum
#of squared residuals does not satisfy the condition abs(fp-s)/s < tol.
#  warnings.warn(message)

